src/main/kotlin/org/rust/cargo/project/workspace/impl/CargoProjectWorkspaceImpl.kt

```diff
@@ -138,54 +154,102 @@ class CargoProjectWorkspaceImpl(private val module: Module) : CargoProjectWorksp
                     }
                 })
 
-                Result.Ok(description)
+                UpdateResult.Ok(description)
             } catch (e: ExecutionException) {
-                Result.Err(e)
+                UpdateResult.Err(e)
             }
         }
 
         override fun onSuccess() {
-            val result = requireNotNull(result)
+            val r = requireNotNull(result)
 
-            when (result) {
-                is Result.Err -> {
-                    showBalloon("Project '${module.project.name}' update failed: ${result.error.message}", MessageType.ERROR)
-                    LOG.info("Project '${module.project.name}' update failed", result.error)
+            lock (lock) {
+                cached = when (r) {
+                    is UpdateResult.Ok -> r.projectDescription
+                    else               -> null
                 }
-                is Result.Ok  ->
+
+                notifyCargoProjectUpdate(r)
+            }
+
+            when (r) {
+                is UpdateResult.Ok  -> LOG.info("Project '${module.project.name}' successfully updated")
+                is UpdateResult.Err -> LOG.info("Project '${module.project.name}' update failed", r.error)
+            }
+        }
+    }
+
+    /**
+     * Project-update task's notifier
+     */
+    inner class Notifier : CargoProjectWorkspaceListener {
+
+        override fun onWorkspaceUpdateCompleted(r: UpdateResult) {
+            when (r) {
+                is UpdateResult.Ok  -> showBalloon("Project '${module.project.name}' successfully updated!", MessageType.INFO)
+                is UpdateResult.Err -> showBalloon("Project '${module.project.name}' update failed: ${r.error.message}", MessageType.ERROR)
+            }
+        }
+
+        private fun showBalloon(message: String, type: MessageType) {
+            PopupUtil.showBalloonForActiveComponent(message, type)
+        }
+    }
+
+    /**
+     * IDEA's project model updater
+     */
+    inner class ProjectModelUpdater : CargoProjectWorkspaceListener {
```

Comment:
Because, single purpose -- single responsibility. `UpdateTask` should do just `cargo metadata` parsing and pushing the conveyor, the rest isn't its responsibility. It should not serve a kitchen combiner purpose. Having things small makes it easily pluggable & maintainable.

No, there is no race-condition. API is low-level & its users may not make any assumptions about the current project-model state.